Remove mongoDB dependency

whitmanschorn opened the pull request:
Removes mongo in favor of postgres 🎉 

## Features changes:

None! Everything should behave exactly as before.

## Testing

Includes updates to all endpoint-based tests, since previous testing relied on mongoDB this iteration has replaced existing mock files with functional mocks defined per-test. 

## Documentation

Now that installation no longer requires mongoDB, any references to it in the docs should be removed and replaces with the relevant PostgreSQL steps.


## Migrations

⚠️ As a first step, ensure that our `Sequence` table is populated properly - it should have one row named `streets` with the same value as our last anonymously created street ID.

Secondly, we run some migrations. Not a schema migration, but rather a SQL command for data re-assignment: Under the current version, this code should merge at the same time as we update the `creatorId` for streets. Currently, a user's `_id` field is used, but we are using `id` as the primary key in most places:

```
UPDATE public."Streets"
SET creator_id = public."Users".id
FROM
public."Users"
WHERE
public."Streets".creator_id = public."Users"._id;
```

A similar script is run for streets involving '_id', here renamed 'legacy_id'. First, we re-import the data from mongo to get legacy IDs:

```
INSERT INTO public."Streets"
SELECT data->>'id' as id, 
  (data->>'namespaced_id')::integer as namespaced_id, 
   (data->>'status')::"enum_Streets_status" as status, 
    data->>'name' as name, 
     (data->'creator_id'->>'$oid') as creator_id, 
      (data->>'data')::json as data,
       (data->'created_at'->>'$date')::timestamp as created_at,
        (data->'updated_at'->>'$date')::timestamp as updated_at,
         (data->'updated_at'->>'$date')::timestamp as client_updated_at,
          data->>'creator_ip' as creator_ip, 
           (data->'original_street_id'->>'$oid') as original_street_id,
		    (data->'_id'->>'$oid') as legacy_id
FROM import.streets
```

Then, we update each original_street_id
```
UPDATE public."Streets" adr1
  SET original_street_id = adr2.id
FROM public."Streets" adr2
WHERE adr1.original_street_id = adr2.legacy_id;
```

louh commented on app/resources/v1/__tests__/user.test.js:
```diff
@@ -1,9 +1,66 @@
 
-jest.mock('../../../models/user')
+jest.mock('twitter')
+jest.mock('../../../db/models', () => {
```
Same comments in this file re:
1. mocks
2. consts

louh commented on app/resources/v1/__tests__/user_session.test.js:
```diff
@@ -3,7 +3,33 @@ import request from 'supertest'
 import session from '../user_session'
 
-jest.mock('../../../models/user')
+jest.mock('../../../db/models', () => {
```
Same comments in this file re:
1. mocks
2. consts

louh commented on app/resources/v1/__tests__/users_pg.test.js:
```diff
@@ -0,0 +1,126 @@
+import users from '../users_pg'
+
+jest.mock('twitter')
+jest.mock('../../../db/models', () => {
```
Same comments in this file re:
1. mocks
2. consts

Also, this file should keep the `users.test.js` filename (no `_pg` suffix).

louh commented on app/resources/v1/street_images.js:
```diff
@@ -128,20 +146,28 @@ exports.post = async function (req, res) {
       res.status(500).json({ status: 500, msg: 'Error finding user.' })
       return
     }
-
```
Put this whitespace back

louh commented on app/resources/v1/user_session.js:
```diff
@@ -1,4 +1,4 @@
+const { User } = require('../../db/models')
 const logger = require('../../../lib/logger.js')()
 
 exports.delete = async function (req, res) {
```
See comment above about retaining use of this API endpoint over exporting the `logout` "verb" from users.

louh commented on app/resources/v1/streets_pg.js:
```diff
@@ -9,14 +9,9 @@ const Op = Sequelize.Op
 
```
TODO: Remove `_pg` suffix from this file.

louh commented on app/resources/v1/streets_pg.js:
```diff
@@ -9,14 +9,9 @@ const Op = Sequelize.Op
-  street.id = hasPriorCall ? res.mainData.id : uuid.v1()
+  const street = {}
+
+  street.id = uuid.v1()
```
Looks like we need to standardize on importing `uuidv1` from `uuid/v1` throughout.

louh commented on app/resources/v1/streets_pg.js:
```diff
@@ -65,8 +60,8 @@ exports.post = async function (req, res) {
-  function updateUserLastStreetId (userId) {
-    return User.findByPk(userId).then((user) => {
+  function updateUserLastStreetId (userAltId) {
+    return User.findOne({ where: { _id: userAltId } }).then((user) => {
```
Let's try to do this without using the `_id` field. See comment above regarding where this comes from.

louh commented on app/resources/v1/users_pg.js:
```diff
@@ -1,5 +1,5 @@
 const config = require('config')
```
TODO: Remove _pg suffix from this file.

louh commented on app/resources/v1/users_pg.js:
```diff
@@ -10,7 +10,7 @@ const Op = Sequelize.Op
 exports.post = async function (req, res) {
   let loginToken = null
-  const hasPriorCall = !!(res.mainData && res.mainData.id)
+  const hasPriorCall = false
```
If this is always false and isn't based on anything anymore, why keep this variable?

louh commented on app/resources/v1/users_pg.js:
```diff
@@ -68,59 +68,28 @@ exports.post = async function (req, res) {
-        } else {
-          user.login_tokens = [loginToken]
+        try {
+          // await User.update(user, { where: { id: credentials.screenName } })
```
Do we need to keep this commented out line? If so, please leave a comment explaining why.

louh commented on app/resources/v1/users_pg.js:
```diff
@@ -68,59 +68,28 @@ exports.post = async function (req, res) {
       res.status(500).json({
         status: 500,
-        msg: 'Error finding user with twitter screenName.'
+        msg: 'Error finding user with Auth Twitter Sign-in.'
```
Replace `Auth` with `Auth0`
Lower case `sign-in`.

louh commented on app/resources/v1/users_pg.js:
```diff
@@ -464,3 +448,43 @@ exports.put = async function (req, res) {
     })
 } // END function - exports.put
+
+exports.logout = async function (req, res) {
```
Please see above comment about not introducing a `logout` verb.

louh commented on app/resources/v1/users_streets_pg.js:
```diff
@@ -2,6 +2,37 @@ const { User, Street } = require('../../db/models')
 const { ERRORS } = require('../../../lib/util')
```
TODO: Remove _pg suffix from this file.

louh commented on app/resources/v1/users_streets_pg.js:
```diff
@@ -30,8 +58,11 @@ exports.get = async function (req, res) {
+    res
+      .status(200)
+      .json(json)
+      .end()
```
You won't need to explicitly `.end()` with a `.json()` response.

louh commented on app/resources/v1/users_streets_pg.js:
```diff
@@ -71,8 +107,76 @@ exports.get = async function (req, res) {
+    const streets = await findUserStreets(user.id)
+    handleFindUserStreets(streets)
+  } catch (err) {
+    console.error(err)
```
If we need to log this, use `logger.error(error)`

louh commented on docs/technical/installing-streetmix.rst:
```diff
@@ -139,13 +100,7 @@ You may already have some of these prerequisites installed. Skip or update the p
-6. Add MongoDB binaries to your system path. Open the Start Menu and type in "environment variables", and select :guilabel:`Edit the system environment variables`. You should see the :guilabel:`Advanced` tab of :guilabel:`System Properties`. Click :guilabel:`Environment Variables...` at the lower right corner of the panel. In the user variables, select or create a variable called ``Path``, then e ...
-
-7. **Optional: Install PostgreSQL.** We are currently working to migrate our database from MongoDB to Postgres. You can `download Windows packages here <https://www.postgresql.org/download/windows/>`.
+4. Install PostgreSQL.You can `download Windows packages here <https://www.postgresql.org/download/windows/>`.
```
A link needs a trailing underscore after the closing backtick.

louh commented on lib/request_handlers/metatags.js:
```diff
@@ -24,7 +23,7 @@ module.exports = async function (req, res, next) {
 
     try {
-      user = await User.findOne({ id: userId })
+      user = await User.findOne({ where: { _id: userId } })
```
Once again, let's do this without using the `_id` field.

louh commented on lib/request_handlers/metatags.js:
```diff
@@ -33,11 +32,15 @@ module.exports = async function (req, res, next) {
 
-    return Street.findOne({ creator_id: user._id, namespaced_id: namespacedId })
+    return Street.findOne({
+      where: { creator_id: user._id, namespaced_id: namespacedId }
```
Let's do this without using the `_id` field. I realize it was used previously, but this shouldn't have happened.

whitmanschorn commented:
thanks for the review @louh, here is a quick update on what I have left to fix:

- Using a shared module for mocking out sequelize fixtures. 
- Correct the user names in the streets gallery (currently shows all streets as owned by 'Anonymous')

whitmanschorn commented:
@louh made all changes. Also did a fair amount of manual testing, so we should be good. Current version has been manually deployed to staging.

louh commented on app/api_routes.js:
```diff
@@ -435,7 +414,7 @@ routes.delete(
   '/api/v1/users/:user_id/login-token',
   cors(),
-  resources.v1.user_session.delete
+  resources.v1.users.logout
```
@whitmanschorn Can you tell me more about why this can't use `user_session.delete`?

louh commented on app/controllers/auth0_sign_in_callback.js:
```diff
@@ -16,23 +16,28 @@ const AccessTokenHandler = function (req, res) {
           const body = response.data
-          res.cookie('user_id', body.id)
+
+          // TODO add code to properly resolve user_id
```
Can you describe the context for the TODO in fuller detail in the comment? What's happening with `user_id` that isn't resolved properly?

whitmanschorn replied:
I've updated this, it now says `// TODO resolve user via auth0`, let me know if you'd like me to remove it. We can set a cookie here for the auth0 library to pick up, that re-write is coming soon 👀

louh commented on test/fixtures/street.json:
```diff
@@ -27,8 +27,8 @@
+  "updated_at": "2018-04-27T20:47:03.477Z",
   "originalStreetId": "33138040-df29-11e9-8f01-c72511055ed4",
   "creator": { "id": "foo" },
   "creatorId": "foo"
```
Does this fixture also need to update the `creatorId` property to `creator_id`?

louh commented on assets/scripts/gallery/GalleryContents.jsx:
```diff
@@ -117,7 +116,10 @@ function GalleryContents (props) {
-              showStreetOwner={userId !== item.creatorId}
+              showStreetOwner={
+                !userId ||
+                !(userId === item.creatorId || userId === item.creator_id)
```
If we're changing `creatorId` to `creator_id` in what cases will `creatorId` still be the property that is present?

whitmanschorn replied:
Have resolved, `creatorId`(and snakeCase in general) will be used everywhere in code.

louh commented on assets/scripts/app/WelcomePanel/FirstTimeExistingStreet.jsx:
```diff
@@ -27,7 +27,7 @@ function FirstTimeExistingStreet (props) {
       <div className="paragraph">
         {/* Display street creator if creatorId is available. */}
-        {(street.creatorId) ? (
+        {street.creatorId ? (
```
Does this need to be updated to `creator_id`?

louh commented on app/resources/v1/users.js:
```diff
@@ -135,44 +141,74 @@ exports.post = function (req, res) {
+        if (!numOfUser) {
+          const newUserData = {
             id: credentials.nickname,
+            _id: credentials.auth0_id.split('|')[1],
```
Are we using this `_id` anywhere?

louh commented on app/resources/v1/users.js:
```diff
@@ -135,44 +141,74 @@ exports.post = function (req, res) {
-          const u = new User({
+          const newUserData = {
             id: id,
+            _id: credentials.auth0_id.split('|')[1],
```
This `_id` as well

louh commented on app/resources/v1/users.js:
```diff
@@ -20,90 +23,90 @@ exports.post = function (req, res) {
+              `Updated data for ${numUsersUpdated} users based on auth0 credentials`
+            )
+          }
+          console.log('user updated!!!')
```
Remove placeholder `console.log`

louh commented on app/resources/v1/users.js:
```diff
@@ -20,90 +23,90 @@ exports.post = function (req, res) {
+          )
+
+          if (numUsersUpdated !== 1) {
+            console.log('!!!!', { numUsersUpdated })
```
Remove placeholder `console.log`

louh commented on app/resources/v1/users.js:
```diff
@@ -20,90 +23,90 @@ exports.post = function (req, res) {
-          access_token_key: twitterCredentials.oauthAccessTokenKey,
-          access_token_secret: twitterCredentials.oauthAccessTokenSecret
+        try {
+          console.log('about to update user in auth0 twitter sign in!!!')
```
Remove placeholder `console.log`

louh commented on app/resources/v1/users.js:
```diff
@@ -20,90 +23,90 @@ exports.post = function (req, res) {
+          const newArray = userUpdates.login_tokens.concat(loginToken)
+          userUpdates.login_tokens = newArray
+        } else {
+          console.log('--- first token')
```
Remove placeholder `console.log`

louh commented on app/resources/v1/users.js:
```diff
@@ -20,90 +23,90 @@ exports.post = function (req, res) {
+        userUpdates.auth0_id = credentials.auth0_id
+        userUpdates.profile_image_url = credentials.profile_image_url
+        if (userUpdates.login_tokens) {
+          console.log('--- adding token')
```
Remove placeholder `console.log`

louh commented on app/resources/v1/users.js:
```diff
@@ -20,90 +23,90 @@ exports.post = function (req, res) {
-        const u = new User({
+        const newUserData = {
           id: credentials.screenName,
+          _id: credentials.auth0_id.split('|')[1],
```
Are we using this `_id` anywhere?

louh commented on app/resources/v1/users.js:
```diff
@@ -20,90 +23,90 @@ exports.post = function (req, res) {
+  }
 
+  const handleUpdateUser = function (user) {
+    // console.log('handleUpdateUser wha?', user );
```
Remove placeholder `console.log`

louh commented on app/resources/v1/user_session.js:
```diff
@@ -13,9 +13,10 @@ exports.delete = async function (req, res) {
+    user = await User.findOne({ where: { id: userId } })
   } catch (err) {
     logger.error(err)
+    console.log('DOOR 1')
```
Remove placeholder `console.log`

louh commented on app/resources/v1/streets.js:
```diff
@@ -162,7 +187,10 @@ exports.delete = async function (req, res) {
-      user = await User.findOne({ login_tokens: { $in: [req.loginToken] } })
+      user = await User.findOne({
+        where: { login_tokens: { $contains: [req.loginToken] } }
+        // where: { login_tokens: { [Op.contains]: [req.loginToken] } }
```
Remove commented out line?

louh replied:
What is the difference between using `$contains` and `[Op.contains]`? Why is it using `[Op.contains]` in line 154 (above)?

louh commented on app/resources/v1/users.js:
```diff
@@ -195,90 +230,288 @@ exports.post = function (req, res) {
+exports.get = async function (req, res) {
+  // Flag error if user ID is not provided
+  const userId = req.params.user_id
+  // console.log('get', { userId })
```
Remove comments if they no longer need to be kept around

louh commented on app/resources/v1/users.js:
```diff
@@ -195,90 +230,288 @@ exports.post = function (req, res) {
+  } // END function - handleFindUser
+
+  const handleError = function (error) {
+    console.log('handleError???', error)
```
Remove stray `console.log`

louh commented on app/resources/v1/users.js:
```diff
@@ -195,90 +230,288 @@ exports.post = function (req, res) {
+    // if (user.login_tokens.indexOf(req.loginToken) === -1) {
+    //   res.status(401).end()
+    //   return
+    // }
```
Why would returning 401 be disabled?

whitmanschorn replied:
This feature gets used when we look at another user's street gallery, for example. Do we want to disable that? I think this was originally written in `users_pg` under that impression but it doesn't seem to be how we're working now.

louh commented on app/resources/v1/users.js:
```diff
@@ -195,90 +230,288 @@ exports.post = function (req, res) {
+    })
+} // END function - exports.put
+
+exports.logout = async function (req, res) {
```
Is this duplicating functionality with `user_sessions.delete`? Why is this here? Why not use `user_sessions.delete`?

louh commented:
@whitmanschorn The revert to package-lock.json is causing CI to fail, can you investigate this please?

louh commented:
@whitmanschorn There is one more bug I've found:

When a logged-out user is editing a street that also has no other user, it's supposed to fork that street and assign it a new ID number. So, for instance, if I were not logged in, and I'm editing https://streetmix.net/-/10, it will create a copy of the street and give it the one plus the last non-user ID number, which at this writing is in `11000000`+.

On this branch, instead of using the last number, it increments by one and then overwrites whatever was there already.

To duplicate this bug:

1. The server should have X number of "anonymous" (no user name attached) streets in the database.
2. Make sure you are logged out.
3. Choose an anonymously-owned street at id Y  (e.g. https://streetmix.net/-/Y) where Y and Y+1 is less than X.
4. Look at the street at /-/Y+1 and remember its contents.
5. Load the street at /-/Y.
6. Make one edit on the street at /-/Y.
7. The URL will change to be /-/Y+1, with your new change. The expected behavior is to change the URL to be /-/X+1.
8. Reload /-/Y+1. The street will have overwritten the original street, which you should have remembered in step 4.

This is a huge problem if we push this live because this means anybody can overwrite streets that haven't been "claimed," and anyone who links to those streets can find unexpected content.

One thing to note (which might help to nail down the source of the bug) is that when I try to do this multiple times, the overwritten street happens at /-/Y+2, /-/Y+3, and so on and so forth; regardless of whether I continue to edit the street at Y or some other ID. It's like the internal counter for "last street" is somehow reset when MongoDB is removed.

whitmanschorn commented:
@louh ah, this makes sense, we keep a dedicated table with a single row for this purpose `Sequences`. I have updated the value on staging and added it as a checklist item for pushing to production.
